docs/_api/compositecontroller.md

```diff
@@ -497,3 +498,63 @@ a chance to recheck the external state without holding up a slot in the work
 queue.
 If you return success, Metacontroller won't call your hook again until the next
 [periodic resync](#resync-period).
+
+### Related Hook
+
+If the `related` hook is defined, Metacontroller will ask for which related
+objects, or classes of objects that your `sync` and `finalize` hooks need to
+know about.
+
+This is useful for mapping across many objects. One example would be a
+controller that lets you specify ConfigMaps to be placed in every Namespace.
+
+Another use-case is being able to reference other objects, e.g. the `env`
+section from a core `Pod` object.
+
+If you don't define a `related` hook, then the related section of the hooks will
+be empty.
+
+The `related` hook will not provide any information about the current state of
+the cluster. Thus, the set of related objects may only depend on the state of
+the parent object.
+
+#### Related Hook Request
+
+A separate request will be sent for each parent object,
+so your hook only needs to think about one parent at a time.
+
+The body of the request (a POST in the case of a [webhook][])
+will be a JSON object with the following fields:
+
+| Field | Description |
+| ----- | ----------- |
+| `controller` | The whole CompositeController object, like what you might get from `kubectl get compositecontroller <name> -o json`. |
+| `parent` | The parent object, like what you might get from `kubectl get <parent-resource> <parent-name> -o json`. |
+
+#### Related Hook Response
+
+The body of your response should be a JSON object with the following fields:
+
+| Field | Description |
+| ----- | ----------- |
+| `labelSelectors` | A list of JSON objects representing all the desired label selectors for this parent object. |
```

Comment:
I would prefer to invert this, so that `labelSelector` is one optional field within a larger object similar to [ResourceRule](https://sourcegraph.com/github.com/GoogleCloudPlatform/metacontroller@353b7d6e7168d478e5fe0aecc716ced192bd92d7/-/blob/apis/metacontroller/v1alpha1/types.go#L47-50). I think that would help make it feel similar to the `childResources` in CompositeController, except this list you get to decide at runtime, whereas that one has to be the same for all parents.

So the hook response could have a field like `relatedResources` which is a list of RelatedResourceRules, which anonymously embed ResourceRule just like [XyzParentResourceRule](https://sourcegraph.com/github.com/GoogleCloudPlatform/metacontroller@353b7d6e7168d478e5fe0aecc716ced192bd92d7/-/blob/apis/metacontroller/v1alpha1/types.go#L52-55), and then also adds `labelSelector` and `namespace`.

Reply:
Sure, have done this